### DISCUSSION TITLE
Performance Concerns with App Router Compared to Pages Router

### DISCUSSION THREAD

**Discussion by buyaoaoman** [vercel/next.js#67048] [App Router]
As Next.js strongly recommends using the App Router, I assumed it would be superior to the Pages Router. However, after a period of development using the App Router, it has become clear that this is not entirely the case. Despite the advanced technologies like RSC and streaming, the response time of the App Router is significantly slower than that of the Pages Router for an SSR framework.

You don’t need a complex example to see this discrepancy. Simply create two new projects using npx create-next-app@latest for both routers, and you’ll notice a huge difference in TTFB (Time to First Byte). The App Router takes almost twice as long as the Pages Router, which is completely unacceptable.

I made this comparison because I initially aimed to enhance user experience by refactoring and upgrading my project from the Pages Router to the App Router. This result is extremely disappointing.

I’m wondering if the current Next.js team is focusing on this issue. If this continues, I might consider switching to Remix or even Nuxt.

**Comment by icyJoseph** [comment]
Got some code? In an app we have, TTFB went down, but interestingly enough, when the app sits behind a CDN that buffers responses, then the TTFB shoots up... expectedly of course, but I wanted to set this up, as a what it was for us, a discovery journey.

For testing sake, I made a create-next-app, with the same contents in pages and app routes:

- [Web Vitals Extension] TTFB 0.012 sec (good) (App Router)
- [Web Vitals Extension] TTFB 0.015 sec (good) (Pages Router)

Refreshing the page several times, gives different results of course, but they two stacks are the same ballpark.

**Reply by buyaoaoman** [reply]
Thank you for your reply and for sharing your test results.

This is my device information:
```Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.5.0: Wed May  1 20:12:58 PDT 2024; root:xnu-10063.121.3~5/RELEASE_ARM64_T6000
  Available memory (MB): 16384
  Available CPU cores: 10
Binaries:
  Node: 18.18.2
  npm: 9.8.1
  Yarn: 1.22.22
  pnpm: 9.1.4
Relevant Packages:
  next: 14.2.4 // Latest available version is detected (14.2.4).
  eslint-config-next: 14.0.4
  react: 18.3.1
  react-dom: 18.3.1
  typescript: 5.1.6
Next.js Config:
  output: N/A
  ```

I tried to implement a simple example by encapsulating a Test component and then rendering it in bulk. The results met my expectations.
```
import styles from "./page.module.css";
import Test from "./conponents/Test";

export default function Home() {
  const array = new Array(200).fill(0);
  return (
    <main className={styles.main}>
      {array.map((_, i) => {
        return <Test key={i} />;
      })}
    </main>
  );
}
```

Test.tsx
```
import Image from "next/image";
import styles from "@/styles/Home.module.css";

export default function Test() {
  return (
    <div>
      <div className={styles.description}>
        <p>
          Get started by editing&nbsp;
          <code className={styles.code}>src/pages/page.tsx</code>
        </p>
        <div>
          <a
            href="https://vercel.com?utm_source=create-next-app&utm_medium=appdir-template&utm_campaign=create-next-app"
            target="_blank"
            rel="noopener noreferrer"
          >
            By{" "}
          </a>
        </div>
      </div>

      <div className={styles.center}></div>

      <div className={styles.grid}>
        <a
          href="https://nextjs.org/docs?utm_source=create-next-app&utm_medium=appdir-template&utm_campaign=create-next-app"
          className={styles.card}
          target="_blank"
          rel="noopener noreferrer"
        >
          <h2>
            Docs <span>-&gt;</span>
          </h2>
          <p>Find in-depth information about Next.js features and API.</p>
        </a>

        <a
          href="https://nextjs.org/learn?utm_source=create-next-app&utm_medium=appdir-template&utm_campaign=create-next-app"
          className={styles.card}
          target="_blank"
          rel="noopener noreferrer"
        >
          <h2>
            Learn <span>-&gt;</span>
          </h2>
          <p>Learn about Next.js in an interactive course with&nbsp;quizzes!</p>
        </a>

        <a
          href="https://vercel.com/templates?framework=next.js&utm_source=create-next-app&utm_medium=appdir-template&utm_campaign=create-next-app"
          className={styles.card}
          target="_blank"
          rel="noopener noreferrer"
        >
          <h2>
            Templates <span>-&gt;</span>
          </h2>
          <p>Explore starter templates for Next.js.</p>
        </a>

        <a
          href="https://vercel.com/new?utm_source=create-next-app&utm_medium=appdir-template&utm_campaign=create-next-app"
          className={styles.card}
          target="_blank"
          rel="noopener noreferrer"
        >
          <h2>
            Deploy <span>-&gt;</span>
          </h2>
          <p>
            Instantly deploy your Next.js site to a shareable URL with Vercel.
          </p>
        </a>
      </div>
    </div>
  );
}
```

In multiple tests, the Pages Router’s TTFB could be as fast as 30+ ms, while the App Router’s fastest TTFB was only 60+ ms. This measurement is purely for server-side rendering speed with the App Router. I mention this because, during the implementation of the example, I found that if I add `'use client'` in the Test component, the App Router becomes faster than the Pages Router. This is likely because the content is shifted to the client-side request in page.js. Additionally, when the Test component includes Next.js’s Image component, the Pages Router becomes much slower, which confuses me a lot.

In my actual project, which contains some server-side components and a large number of client-side components, the final result is that in the local development environment, the App Router is about 100 ms slower than the Pages Router. However, in the production environment, the difference is 300-400 ms, which is completely unacceptable to me. I have read the Next.js documentation multiple times and explored the community extensively, but I still haven’t found an answer.

**Reply by icyJoseph** [reply]
Yeah this kind of example has been observed before, https://github.com/eknkc/ssr-benchmark

I haven't had time to debug this much more than this though:

This image `/landing`, is a Pages Router, next build, statically optimized page:

<img width="1438" alt="Screenshot 2024-06-20 at 11 46 05" src="https://github.com/vercel/next.js/assets/21013447/7e711d15-48e1-4b10-880d-9f40275500d8">

This is an App Router, next build, statically optimized page:

<img width="1438" alt="Screenshot 2024-06-20 at 11 45 54" src="https://github.com/vercel/next.js/assets/21013447/af12b898-abb2-49e6-8a21-4d8d7c04d267">

Here's the build report for transparency:

```console
Route (app)                               Size     First Load JS
┌ ○ /                                     136 B          87.2 kB
└ ○ /_not-found                           875 B          87.9 kB
+ First Load JS shared by all             87.1 kB
  ├ chunks/23-ef3c75ca91144cad.js         31.5 kB
  ├ chunks/fd9d1056-2821b0f0cabcd8bd.js   53.7 kB
  └ other shared chunks (total)           1.88 kB

Route (pages)                             Size     First Load JS
┌   /_app                                 0 B            79.1 kB
└ ○ /landing                              727 B          79.9 kB
+ First Load JS shared by all             80.7 kB
  ├ chunks/framework-4be839806aa8e2d3.js  45.2 kB
  ├ chunks/main-66783b0b0ef1a702.js       32 kB
  └ other shared chunks (total)           3.5 kB

○  (Static)  prerendered as static content
```

As you can see, in the second image there's a lot of "dead time" to the right, after all traced work seems to have completed, I guess this is either, a bunch of promises resolving, or a stream not completing? But yeah, that one use case is strange to say the least, and although not very likely in real world apps, it can still happen.

**Reply by buyaoaoman** [reply]
ye I konw this https://github.com/eknkc/ssr-benchmark.

I am raising this discussion to find out if the Next.js team is actively working on optimizing this issue and when it might be resolved. As a developer using the framework, if I am expected to transition from Pages Router to App Router, I believe there shouldn’t be such a significant gap in basic server-side rendering performance. TTFB is still a very important metric despite the advantages of streaming and RSC (is RSC still an advantage in this context?). Overall, the page’s FCP remains very low.

I am very grateful to the Next.js team for their contributions to open source, and I hope that the App Router can be improved. Please forgive me for pointing out flaws from a user’s perspective, but my technical knowledge and time do not allow me to dive into the source code to address this myself.

Any insights or updates on this matter would be greatly appreciated.

**Reply by icyJoseph** [reply]
Streaming is supposed to bring down TTFB, because the server sends data ASAP, without blocking, but that has caveats, gotchas, and other bits.

I think, unfortunately, that people just look at metrics raw, and use umbrella terms all over the place. 

In this case though, it is indeed, at the very least, weird, that the request has so much dead time to the right of the last finished span. 

Take for instance, what if we make everything render on demand:

This time the first image is Pages Router, with a getServerSideProps function that does nothing:

<img width="1459" alt="Screenshot 2024-06-20 at 12 29 19" src="https://github.com/vercel/next.js/assets/21013447/62112fd9-e28a-4764-aa51-a54a92bc6963">

This second image is App Router with `force-dynamic`:

<img width="1459" alt="Screenshot 2024-06-20 at 12 32 41" src="https://github.com/vercel/next.js/assets/21013447/d49a8cd9-50b0-45e9-9d05-b6e6d778a165">

Both using the same setup as before. Build summary here for transparency:

```console
Route (app)                               Size     First Load JS
┌ ƒ /                                     136 B          87.2 kB
└ ○ /_not-found                           875 B          87.9 kB
+ First Load JS shared by all             87.1 kB
  ├ chunks/23-ef3c75ca91144cad.js         31.5 kB
  ├ chunks/fd9d1056-2821b0f0cabcd8bd.js   53.7 kB
  └ other shared chunks (total)           1.88 kB

Route (pages)                             Size     First Load JS
┌   /_app                                 0 B            79.1 kB
└ ƒ /landing                              743 B          79.9 kB
+ First Load JS shared by all             80.7 kB
  ├ chunks/framework-4be839806aa8e2d3.js  45.2 kB
  ├ chunks/main-66783b0b0ef1a702.js       32 kB
  └ other shared chunks (total)           3.5 kB

○  (Static)   prerendered as static content
ƒ  (Dynamic)  server-rendered on demand
```

But now App Router does better, though TTFB from webvitals, shows them at essentially the same level:

- Pages Router, [Web Vitals Extension] TTFB 0.059 sec (good)
- App Router, [Web Vitals Extension] TTFB 0.069 sec (good)

And, you can also get better results over HTTPS, on your former example, IIRC. 

I guess, what I am trying to get at, is that, this kind of measurement, is not necessary very easy to do, because it depends on a whole bunch of factors.

**Reply by SalmenBejaoui** [reply]
@icyJoseph are those screenshot generated by this repo https://github.com/eknkc/ssr-benchmark? I would like to benchmark my app like this. Thank you

**Reply by buyaoaoman** [reply]
@icyJoseph You’re right, we shouldn’t just focus on certain metrics. So, is streaming rendering one of the reasons why App Router’s server-side rendering time might be longer than Pages Router’s?

**Reply by icyJoseph** [reply]
You can follow this guide: https://nextjs.org/docs/app/building-your-application/optimizing/open-telemetry

- enable the instrumentation hook
- in the instrumentation file, import and use @vercel/otel (also install it)
- in your .env add NEXT_OTEL_VERBOSE=1
- then go to this repo https://github.com/vercel/opentelemetry-collector-dev-setup, clone and run the recommended command
- now build and start your app, and go to the zipkin consumer, and find your service name, `next-app` if you followed the guide, there 😄 

I'd also recommend using HTTPS, with Caddy you can reverse-proxy your localhost to your next.js server, `caddy reverse-proxy --to :3000`, so that you can go to `https://localhost` (no port), and see your app. That'll make it pick up http2 streaming, and to verify this, you can install the Chrome extension Claire, https://chromewebstore.google.com/detail/claire/fgbpcgddpmjmamlibbaobboigaijnmkl

You can also use `oha`, https://github.com/hatoo/oha, to fire a bunch of requests to your server, and see how it behaves:

```console
oha http://localhost:3006 --http-version=1.1

Response time histogram:
  0.048 [1]   |
  0.217 [113] |■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.387 [48]  |■■■■■■■■■■■■■
  0.556 [4]   |■
  0.726 [3]   |
  0.895 [2]   |
  1.065 [3]   |
  1.234 [2]   |
  1.404 [2]   |
  1.573 [2]   |
  1.743 [20]  |■■■■■
```

but using HTTPS with HTTP2, through caddy:

```console
oha https://localhost --http-version=2

Response time histogram:
  0.172 [1]  |
  0.279 [14] |■■■■
  0.386 [97] |■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.494 [78] |■■■■■■■■■■■■■■■■■■■■■■■■■
  0.601 [0]  |
  0.708 [1]  |
  0.815 [2]  |
  0.923 [1]  |
  1.030 [2]  |
  1.137 [1]  |
  1.244 [3]  |
```

Both are firing request to the same App Router page, statically rendered during next build, using next start.

**Reply by buyaoaoman** [reply]
I used to think that App Router would achieve faster initial response times and thus a faster TTFB by using streaming. However, pure server-side rendering doesn’t seem to yield that result, which is quite surprising.

**Reply by luisvieiragmr** [reply]
> Yeah this kind of example has been observed before, https://github.com/eknkc/ssr-benchmark
> 
> I haven't had time to debug this much more than this though:
> 
> This image `/landing`, is a Pages Router, next build, statically optimized page:
> 
> <img alt="Screenshot 2024-06-20 at 11 46 05" width="1438" src="https://private-user-images.githubusercontent.com/21013447/341364011-7e711d15-48e1-4b10-880d-9f40275500d8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzcxMjQwMDYsIm5iZiI6MTczNzEyMzcwNiwicGF0aCI6Ii8yMTAxMzQ0Ny8zNDEzNjQwMTEtN2U3MTFkMTUtNDhlMS00YjEwLTg4MGQtOWY0MDI3NTUwMGQ4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTE3VDE0MjE0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkwZTVjZjJmMjc1MGE1MjExM2YwZDlkNDBhZThjMDJmZGIyNmJhN2JhYmQ0NmI1NTk3MjMwMWVmYjJjZWY2MDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.aO29zvVAFteSIFZS2xnXGxXnhB3SUm91ZTam_5aI2yM">
> This is an App Router, next build, statically optimized page:
> 
> <img alt="Screenshot 2024-06-20 at 11 45 54" width="1438" src="https://private-user-images.githubusercontent.com/21013447/341364003-af12b898-abb2-49e6-8a21-4d8d7c04d267.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzcxMjQwMDYsIm5iZiI6MTczNzEyMzcwNiwicGF0aCI6Ii8yMTAxMzQ0Ny8zNDEzNjQwMDMtYWYxMmI4OTgtYWJiMi00OWU2LThhMjEtNGQ4ZDdjMDRkMjY3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTE3VDE0MjE0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE3ODk1YjBmZDliODg0M2VhYzM4NjFkNTRkYjRiNDc2Y2JkOTM2NWUxN2Q0Y2Y5NGU1MWFlNGJlOTU3N2ExMDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.mQgpy2txmfbnRGGyJUietZzTKurhPfjZSiucU5SLvbU">
> Here's the build report for transparency:
> 
> ```
> Route (app)                               Size     First Load JS
> ┌ ○ /                                     136 B          87.2 kB
> └ ○ /_not-found                           875 B          87.9 kB
> + First Load JS shared by all             87.1 kB
>   ├ chunks/23-ef3c75ca91144cad.js         31.5 kB
>   ├ chunks/fd9d1056-2821b0f0cabcd8bd.js   53.7 kB
>   └ other shared chunks (total)           1.88 kB
> 
> Route (pages)                             Size     First Load JS
> ┌   /_app                                 0 B            79.1 kB
> └ ○ /landing                              727 B          79.9 kB
> + First Load JS shared by all             80.7 kB
>   ├ chunks/framework-4be839806aa8e2d3.js  45.2 kB
>   ├ chunks/main-66783b0b0ef1a702.js       32 kB
>   └ other shared chunks (total)           3.5 kB
> 
> ○  (Static)  prerendered as static content
> ```
> 
> As you can see, in the second image there's a lot of "dead time" to the right, after all traced work seems to have completed, I guess this is either, a bunch of promises resolving, or a stream not completing? But yeah, that one use case is strange to say the least, and although not very likely in real world apps, it can still happen.

Did you ever figure out what this "dead time" is? i'm currently struggling with something similar

**Reply by icyJoseph** [reply]
Not yet; but I will. That being said, i remember doing some other benchmark with version 15, and there were some improvements

**Comment by Envoy49** [comment]
Hey peeps, I opened this issue https://github.com/vercel/next.js/issues/69561
App router is significantly slower than Pages router even in sandbox environment

**Comment by Envoy49** [comment]
https://www.youtube.com/watch?v=3Q2q2gs0nAI

found a video which is relevant to our problems

**Comment by smoothdvd** [comment]
Same situation when I integrate bull-board + hone into NextJS 15.

In Page router mode:
```
wrk -t8 -c64 -d15s --timeout 8s http://localhost:3003/api/bull
Running 15s test @ http://localhost:3003/api/bull
  8 threads and 64 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    12.77ms   38.86ms 657.50ms   98.17%
    Req/Sec     0.98k   113.57     1.32k    91.81%
  116028 requests in 15.04s, 149.38MB read
Requests/sec:   7712.06
Transfer/sec:      9.93MB
```

In App router mode:
```
 wrk -t8 -c64 -d15s --timeout 8s http://localhost:3003/api/queues
Running 15s test @ http://localhost:3003/api/queues
  8 threads and 64 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    24.15ms   36.09ms 720.03ms   98.48%
    Req/Sec   390.59     58.98     1.02k    84.90%
  46526 requests in 15.08s, 63.63MB read
Requests/sec:   3085.63
Transfer/sec:      4.22MB
```

Page router's Req/s is almost 2.5x faster than App router.

**Comment by wirelessjeano** [comment]
This is still an issue in v15.2.0. TTFB is significantly slower in App Router in standalone mode. I released via Vercel and did some testing. Not surprisingly, App Router is rocket fast if you host with Vercel.
